--- a/usr/share/jolla-alarm-ui/pages/AlarmDialogBase.qml
+++ b/usr/share/jolla-alarm-ui/pages/AlarmDialogBase.qml
@@ -13,6 +13,8 @@ SilicaFlickable {
 
     property QtObject alarm: dummy
 
+    property bool dismissMenuVisible: true
+
     property int status: -1
     property alias topIconSource: topIcon.source
     default property alias _content: content.children
@@ -20,6 +22,7 @@ SilicaFlickable {
     property alias spacing: content.spacing
     property alias animating: fadeOut.running
 
+    signal dialogShown()
     signal dialogHidden(int status)
     signal timeout()
 
@@ -33,6 +36,7 @@ SilicaFlickable {
         opacity = 1.0
         feedback.play()
         timeoutTimer.start()
+        dialogShown()
     }
 
     function closeDialog(status) {
@@ -77,6 +81,7 @@ SilicaFlickable {
     }
 
     PulleyAnimationHint {
+        visible: dismissMenuVisible
         id: pulleyAnimationHint
         anchors.fill: parent
         pushUpHint: true
@@ -109,6 +114,7 @@ SilicaFlickable {
     Image {
         id: dismissIcon
 
+        visible: dismissMenuVisible
         anchors {
             bottom: parent.bottom
             bottomMargin: Theme.paddingLarge
@@ -118,6 +124,7 @@ SilicaFlickable {
     }
 
     PushUpMenu {
+        visible: dismissMenuVisible
         quickSelect: true
         Item { height: Theme.itemSizeExtraSmall; width: parent.width }
         MenuItem {
--- a/usr/share/jolla-alarm-ui/pages/ClockAlarmDialog.qml
+++ b/usr/share/jolla-alarm-ui/pages/ClockAlarmDialog.qml
@@ -7,7 +7,54 @@ AlarmDialogBase {
     onTimeout: closeDialog(AlarmDialogStatus.Closed)
 
     topIconSource: "image://theme/icon-l-snooze?" + Theme.highlightColor
+    
+    dismissMenuVisible: false
 
+    property string question
+    property int response
+    property variant operators: ["+", "-", "x"]
+    
+    function setup_question() {
+                
+        var one = Math.floor((Math.random() * 10) + 1);
+        var two = Math.floor((Math.random() * 10) + 1);
+
+        var operator = operators[Math.floor(Math.random() * 4)];
+
+        /* Ensure the first is the highest */
+        if (two > one) {
+            var hey = one;
+            one = two;
+            two = hey;
+        }
+
+        /*
+         * In some strange cases, the operator is undefined.
+         * Force it to be +.
+        */
+        (operator == undefined) ? operator = "+" : operator;
+
+        switch (operator) {
+
+        case "+":
+            response = one + two;
+            break;
+        case "-":
+            response = one - two;
+            break;
+        case "x":
+            response = one * two;
+            break;
+
+        }
+
+        question = "%1 %2 %3".arg(one).arg(operator).arg(two);
+        mathsField.focus = true;
+
+    }
+    
+    onDialogShown: setup_question()
+    
     PullDownMenu {
         quickSelect: true
         MenuItem {
@@ -71,4 +118,19 @@ AlarmDialogBase {
         text: alarm.title
         wrapMode: Text.Wrap
     }
+
+    TextField {
+
+        id: mathsField
+        width: parent.width
+        placeholderText: question
+        label: question
+        inputMethodHints: Qt.ImhDigitsOnly
+        onTextChanged: {
+            if (text == response) {
+                closeDialog(AlarmDialogStatus.Dismissed);
+            }
+        }
+    }
+
 }
